fl4health/clients/flash_client.py

```diff
@@ -71,69 +79,22 @@ def train_by_epochs(
 
             metrics = self.train_metric_manager.compute()
             loss_dict = self.train_loss_meter.compute().as_dict()
-            current_loss = loss_dict.get("backward", 0.0)
-
-            # Early stopping check
-            if self.gamma is not None and abs(previous_loss - current_loss) < self.gamma / (local_step + 1):
-                log(
-                    INFO, f"Early stopping at epoch {local_epoch} with loss change {abs(previous_loss - current_loss)}"
-                )
-                break
-
-            previous_loss = current_loss
+            current_loss, _ = self.validate()
 
             self._handle_logging(loss_dict, metrics, current_round=current_round, current_epoch=local_epoch)
             self._handle_reporting(loss_dict, metrics, current_round=current_round)
 
-        return loss_dict, metrics
-
-    def train_by_steps(
-        self, steps: int, current_round: Optional[int] = None
-    ) -> Tuple[Dict[str, float], Dict[str, Scalar]]:
-        self.model.train()
-
-        # Pass loader to iterator so we can step through train loader
-        train_iterator = iter(self.train_loader)
-
-        self.train_loss_meter.clear()
-        self.train_metric_manager.clear()
-        previous_loss = float("inf")
-        for step in range(steps):
-            try:
-                input, target = next(train_iterator)
-            except StopIteration:
-                train_iterator = iter(self.train_loader)
-                input, target = next(train_iterator)
-
-            if self.is_empty_batch(input):
-                log(INFO, "Empty batch generated by data loader. Skipping step.")
-                continue
-
-            input, target = self._move_input_data_to_device(input), target.to(self.device)
-            losses, preds = self.train_step(input, target)
-            self.train_loss_meter.update(losses)
-            self.train_metric_manager.update(preds, target)
-            self.update_after_step(step)
-            self.total_steps += 1
-
-            loss_dict = self.train_loss_meter.compute().as_dict()
-            current_loss = loss_dict.get("backward", 0.0)
-
-            # Early stopping check
-            if self.gamma is not None and abs(previous_loss - current_loss) < self.gamma / (step + 1):
-                log(INFO, f"Early stopping at step {step} with loss change {abs(previous_loss - current_loss)}")
+            if self.gamma is not None and previous_loss - current_loss < self.gamma / (local_epoch + 1):
+                log(
+                    INFO, f"Early stopping at epoch {local_epoch} with loss change {abs(previous_loss - current_loss)}"
```

Comment:
I think it's worth logging what gamma is in this message just for visual comparison.